zstd: data race in shared Decoder.DecodeAll when a trained dictionary is used (blockDec.decodeLiterals writes dict.litEnc.MaxDecodedSize)

The `zstd` docs state that a `Decoder` "can be used for concurrent decompression of multiple buffers." This does not hold when the decoder is configured with a **trained** dictionary (`WithDecoderDicts`) and the frames use *treeless* literal blocks: concurrent `DecodeAll` calls on one shared decoder race on the dictionary's shared huffman scratch.

## Root cause

`history.setDict` aliases the shared dictionary's literal encoder into the per-frame history:

```go
// zstd/history.go
func (h *history) setDict(dict *dict) {
    ...
    h.huffTree = dict.litEnc // shared *huff0.Scratch, owned by the dict
}
```

For a **treeless** literals block, `decodeLiterals` uses that shared scratch directly and writes into it:

```go
// zstd/blockdec.go  (case literalsBlockTreeless)
huff := hist.huffTree            // == dict.litEnc (shared across all goroutines)
...
huff.MaxDecodedSize = litRegenSize   // <-- write into shared state (blockdec.go:404)
literals, err = huff.Decoder().Decompress4X(...)
```

Because every frame that uses this dictionary points `hist.huffTree` at the *same* `dict.litEnc`, two goroutines decoding treeless frames on the same shared decoder write `MaxDecodedSize` on the same `*huff0.Scratch` concurrently.

Note the neighboring **compressed** literals case is already defensive — when `huff == hist.dict.litEnc` it swaps in a fresh scratch from `huffDecoderPool` before mutating — so only the treeless path is affected. This is also why the race only shows up with **trained** dictionaries: raw dictionaries (`WithDecoderDictRaw`) carry no `litEnc`, so `hist.huffTree` is nil and the treeless path is never taken from a shared table.

## Reproduce

Self-contained test. The dictionary corpus is a strongly-skewed low-entropy alphabet so that (a) huffman beats raw storage (blocks are emitted Compressed rather than Raw) and (b) the small symbol set lets the encoder reuse the dictionary's trained table across payloads, i.e. emit *treeless* literal blocks — the affected path. (`BuildDict` also requires the raw dict content via `History` and non-zero `Offsets`, otherwise it errors on load.)

```go
package minrepro

import (
	"bytes"
	"math/rand"
	"sync"
	"testing"

	"github.com/klauspost/compress/zstd"
)

func alphabet() []byte {
	w := map[byte]int{'e': 40, 't': 20, 'a': 12, 'o': 8, 'i': 6, 'n': 5, 's': 4, 'h': 3, 'r': 2}
	var b []byte
	for c, n := range w {
		for i := 0; i < n; i++ {
			b = append(b, c)
		}
	}
	return b
}

func sample(rng *rand.Rand, n int, a []byte) []byte {
	b := make([]byte, n)
	for i := range b {
		b[i] = a[rng.Intn(len(a))]
	}
	return b
}

func TestSharedDecoderTrainedDictRace(t *testing.T) {
	a := alphabet()
	rng := rand.New(rand.NewSource(2000))

	var samples [][]byte
	for i := 0; i < 2000; i++ {
		samples = append(samples, sample(rng, 32+rng.Intn(200), a))
	}
	dict, err := zstd.BuildDict(zstd.BuildDictOptions{
		ID:       12345,
		Contents: samples,
		History:  sample(rng, 12<<10, a),
		Offsets:  [3]int{1, 4, 8},
	})
	if err != nil {
		t.Fatal(err)
	}

	enc, _ := zstd.NewWriter(nil, zstd.WithEncoderDict(dict), zstd.WithEncoderConcurrency(1))
	var frames, plains [][]byte
	for i := 0; i < 400; i++ {
		pt := sample(rng, 200+rng.Intn(800), a)
		frames = append(frames, enc.EncodeAll(pt, nil))
		plains = append(plains, pt)
	}
	enc.Close()

	// ONE shared decoder, many goroutines calling DecodeAll concurrently.
	dec, err := zstd.NewReader(nil, zstd.WithDecoderDicts(dict))
	if err != nil {
		t.Fatal(err)
	}
	defer dec.Close()

	var wg sync.WaitGroup
	for g := 0; g < 48; g++ {
		wg.Add(1)
		go func(seed int) {
			defer wg.Done()
			r := rand.New(rand.NewSource(int64(seed)))
			for i := 0; i < 3000; i++ {
				idx := r.Intn(len(frames))
				out, err := dec.DecodeAll(frames[idx], nil)
				if err != nil {
					t.Errorf("decode error: %v", err)
					return
				}
				if !bytes.Equal(out, plains[idx]) {
					t.Errorf("output mismatch")
					return
				}
			}
		}(g + 1)
	}
	wg.Wait()
}
```

```
go test -race -run TestSharedDecoderTrainedDictRace
```

## Result

`go test -race` reports a data race on both the current release and the previous one; a control run with one decoder per goroutine (no sharing) is clean.

`-race` output (v1.19.1, Go 1.26.4 darwin/arm64):

```
WARNING: DATA RACE
Write at 0x00c00496b950 by goroutine 23:
  github.com/klauspost/compress/zstd.(*blockDec).decodeLiterals()
      .../zstd/blockdec.go:404 +0x1538
  github.com/klauspost/compress/zstd.(*blockDec).decodeCompressed()
      .../zstd/blockdec.go:479 +0x48
  github.com/klauspost/compress/zstd.(*blockDec).decodeBuf()
      .../zstd/blockdec.go:256 +0x44c
  github.com/klauspost/compress/zstd.(*frameDec).runDecoder()
      .../zstd/framedec.go:371 +0x2f0
  github.com/klauspost/compress/zstd.(*Decoder).DecodeAll()
      .../zstd/decoder.go:395 +0x7a4
  minrepro.TestSharedDecoderTrainedDictRace.func1()
      .../repro_test.go:81

Previous write at 0x00c00496b950 by goroutine 21:
  github.com/klauspost/compress/zstd.(*blockDec).decodeLiterals()
      .../zstd/blockdec.go:404 +0x1538
  github.com/klauspost/compress/zstd.(*blockDec).decodeCompressed()
      .../zstd/blockdec.go:479 +0x48
  github.com/klauspost/compress/zstd.(*blockDec).decodeBuf()
      .../zstd/blockdec.go:256 +0x44c
  github.com/klauspost/compress/zstd.(*frameDec).runDecoder()
      .../zstd/framedec.go:371 +0x2f0
  github.com/klauspost/compress/zstd.(*Decoder).DecodeAll()
      .../zstd/decoder.go:395 +0x7a4
  minrepro.TestSharedDecoderTrainedDictRace.func1()
      .../repro_test.go:81
==================
--- FAIL: TestSharedDecoderTrainedDictRace ... race detected during execution of test
```

The colliding address is the `MaxDecodedSize` field of the dictionary's shared `*huff0.Scratch` (`blockdec.go:404`, the treeless case). The same race reproduces on **v1.18.0** at the identical line.

Shared vs. pooled contrast, 48 goroutines:

| variant | v1.18.0 | v1.19.1 |
|---|---|---|
| shared decoder, `-race` | DATA RACE at blockdec.go:404 | DATA RACE at blockdec.go:404 |
| one decoder per goroutine, `-race` | clean (PASS) | clean (PASS) |

Without `-race`, at 48 goroutines × 30000 iterations, the shared run happened to produce no output mismatches or errors in these runs — the two racing writes usually store equivalent values — but the concurrent write remains a data race (undefined behavior under the Go memory model) and could surface as a spurious `literal output size mismatch` decode error or corrupted output under other timings.

## Suggested fix

Give the treeless case the same treatment the compressed case already has: when `hist.huffTree == hist.dict.litEnc`, decode against a scratch obtained from `huffDecoderPool` (its `dt` table copied/transferred from the dictionary's) rather than writing `MaxDecodedSize`/decoding on the shared `dict.litEnc`. Alternatively, document that a decoder holding a **trained** dictionary is not safe for concurrent `DecodeAll` and must be pooled per goroutine.

Environment: `github.com/klauspost/compress` v1.18.0 and v1.19.1, Go 1.26.4, darwin/arm64.

--- a/zstd/blockdec.go
+++ b/zstd/blockdec.go
@@ -400,8 +400,9 @@ func (b *blockDec) decodeLiterals(in []byte, hist *history) (remain []byte, err 
 			}
 		}
 		var err error
-		// Use our out buffer.
-		huff.MaxDecodedSize = litRegenSize
+		// Decoder.Decompress* uses cap(dst) for the size limit. Do not write
+		// MaxDecodedSize on hist.huffTree: with a trained dictionary that
+		// pointer aliases the shared dict.litEnc and concurrent DecodeAll races.
 		if fourStreams {
 			literals, err = huff.Decoder().Decompress4X(b.literalBuf[:0:litRegenSize], literals)
 		} else {

--- a/zstd/dict_test.go
+++ b/zstd/dict_test.go
@@ -4,8 +4,10 @@ import (
 	"bytes"
 	"fmt"
 	"io"
+	"math/rand"
 	"os"
 	"strings"
+	"sync"
 	"testing"
 
 	"github.com/klauspost/compress/zip"
@@ -885,4 +887,87 @@ func TestEncoderDictAddViaReset(t *testing.T) {
 			}
 		})
 	}
+}
+
+// TestSharedDecoderTrainedDictRace verifies concurrent DecodeAll on one Decoder
+// that holds a trained dictionary does not race on dict.litEnc (gh-1180).
+func TestSharedDecoderTrainedDictRace(t *testing.T) {
+	weights := map[byte]int{'e': 40, 't': 20, 'a': 12, 'o': 8, 'i': 6, 'n': 5, 's': 4, 'h': 3, 'r': 2}
+	var alphabet []byte
+	for c, n := range weights {
+		for i := 0; i < n; i++ {
+			alphabet = append(alphabet, c)
+		}
+	}
+	sample := func(rng *rand.Rand, n int) []byte {
+		b := make([]byte, n)
+		for i := range b {
+			b[i] = alphabet[rng.Intn(len(alphabet))]
+		}
+		return b
+	}
+
+	rng := rand.New(rand.NewSource(2000))
+	samples := make([][]byte, 2000)
+	for i := range samples {
+		samples[i] = sample(rng, 32+rng.Intn(200))
+	}
+	dict, err := BuildDict(BuildDictOptions{
+		ID:       12345,
+		Contents: samples,
+		History:  sample(rng, 12<<10),
+		Offsets:  [3]int{1, 4, 8},
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	enc, err := NewWriter(nil, WithEncoderDict(dict), WithEncoderConcurrency(1))
+	if err != nil {
+		t.Fatal(err)
+	}
+	nFrames := 400
+	iters := 3000
+	goroutines := 48
+	if testing.Short() {
+		nFrames = 80
+		iters = 200
+		goroutines = 8
+	}
+	frames := make([][]byte, nFrames)
+	plains := make([][]byte, nFrames)
+	for i := 0; i < nFrames; i++ {
+		pt := sample(rng, 200+rng.Intn(800))
+		frames[i] = enc.EncodeAll(pt, nil)
+		plains[i] = pt
+	}
+	enc.Close()
+
+	dec, err := NewReader(nil, WithDecoderDicts(dict))
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer dec.Close()
+
+	var wg sync.WaitGroup
+	for g := 0; g < goroutines; g++ {
+		wg.Add(1)
+		go func(seed int) {
+			defer wg.Done()
+			r := rand.New(rand.NewSource(int64(seed)))
+			for i := 0; i < iters; i++ {
+				idx := r.Intn(len(frames))
+				out, err := dec.DecodeAll(frames[idx], nil)
+				if err != nil {
+					t.Errorf("decode error: %v", err)
+					return
+				}
+				if !bytes.Equal(out, plains[idx]) {
+					t.Errorf("output mismatch")
+					return
+				}
+			}
+		}(g + 1)
+	}
+	wg.Wait()
 }

--- a/zstd/dict_test.go
+++ b/zstd/dict_test.go
@@ -895,7 +895,7 @@ func TestSharedDecoderTrainedDictRace(t *testing.T) {
 	weights := map[byte]int{'e': 40, 't': 20, 'a': 12, 'o': 8, 'i': 6, 'n': 5, 's': 4, 'h': 3, 'r': 2}
 	var alphabet []byte
 	for c, n := range weights {
-		for i := 0; i < n; i++ {
+		for range n {
 			alphabet = append(alphabet, c)
 		}
 	}

